Kayagokalp/test without panic

kayagokalp opened the pull request:
changelog: Adds a `test_without_fail_case` lint which checks if a test can fail or not. If failure is not possible, the test is linted. This is done to prevent any silently failing tests that actually was not asserting anything, improving test suite quality.

kayagokalp commented on README.md:
```diff
@@ -5,7 +5,7 @@
 A collection of lints to catch common mistakes and improve your [Rust](https://github.com/rust-lang/rust) code.
 
-[There are over 700 lints included in this crate!](https://rust-lang.github.io/rust-clippy/master/index.html)
+[There are over 750 lints included in this crate!](https://rust-lang.github.io/rust-clippy/master/index.html)
```
Not sure why this is showing as 750.

kayagokalp commented:
@rustbot review

kayagokalp commented:
@rustbot author

kayagokalp commented:
I am adding configuration capabilities to this PR and will also add more tests.

blyxyas commented on clippy_lints/src/test_without_fail_case.rs:
```diff
@@ -0,0 +1,270 @@
+                    // unlikely that this is intended as the tests assertion. In the name of
+                    // reducing false negatives we are giving out soundness.
+                    //
+                    // This decision can be justified as it is highly unlikely that this lint is sound
+                    // without this additional check, and with this we are reducing the number of false
+                    // negatives.
```
```suggestion
                    // This reduces the number of false positives
```

blyxyas reviewed (changes requested):
Good starting point, but I'm worried for performance and memory usage. I'm trying to think of a way to check if a function links to the `panic` handler DefId.

kayagokalp commented on clippy_lints/src/test_without_fail_case.rs:
```diff
@@ -0,0 +1,270 @@
+    pub fn new(conf: &Conf) -> Self {
+        Self {
+            config: SearchConfig {
+                indexing_fallible: conf.test_without_fail_case_include_indexing_as_fallible,
+                fallible_paths: conf.test_without_fail_case_fallible_paths.iter().cloned().collect(),
+                non_fallible_paths: conf.test_without_fail_case_non_fallible_paths.iter().cloned().collect(),
```
Mostly out of habit of using `FxHashSet` in the compiler for performance reasons. But given that this is only the configuration I'll change it into `HashSet`.

kayagokalp replied:
I guess it is a small lookup table, as the list coming from configuration file likely to be limited, can't see a reason why it couldn't be a `Vec` for sure. Can change it.

kayagokalp commented on book/src/lint_configuration.md:
```diff
@@ -831,6 +831,52 @@ if no suggestion can be made.
+indexing an array is not the intended way to fail a test. So setting this `true`
+reduces false positives but makes the analysis more focused on possible byproducts
+of a test. That is the set of operations to get the point we assert something rather
+than the existance of asserting that thing.
```
> I think it would be excellent to have a way to, instead of adding a third configuration option, be able to use std::ops::Index::index or similar in the other two configuration options.

I agree, I'll think about it

flip1995 commented on CHANGELOG.md:
```diff
@@ -6215,6 +6216,9 @@ Released 2018-09-13
 [`standard-macro-braces`]: https://doc.rust-lang.org/clippy/lint_configuration.html#standard-macro-braces
 [`struct-field-name-threshold`]: https://doc.rust-lang.org/clippy/lint_configuration.html#struct-field-name-threshold
 [`suppress-restriction-lint-in-const`]: https://doc.rust-lang.org/clippy/lint_configuration.html#suppress-restriction-lint-in-const
+[`test-without-fail-case-fallible-paths`]: https://doc.rust-lang.org/clippy/lint_configuration.html#test-without-fail-case-fallible-paths
+[`test-without-fail-case-include-indexing-as-fallible`]: https://doc.rust-lang.org/clippy/lint_configuration.html#test-without-fail-case-include-indexing-as-fallible
+[`test-without-fail-case-non-fallible-paths`]: https://doc.rust-lang.org/clippy/lint_configuration.html#test-without-fail-case-non-fallible-paths
```
I find 3 configuration values for this lint a bit much. Also those are a mouth full. 

I think there should only one configuration option at most: A list of items/functions that should prevent triggering the lint.

kayagokalp commented:
> Good starting point, but I'm worried for performance and memory usage. I'm trying to think of a way to check if a function links to the panic handler DefId.

If there is way to see if a function links to a panic handler without visiting them explicitly that would be great! Otherwise to access the performance characteristics it might be wise to take a look at set of benchmarks. I am curious to see how many interleaved levels of functions calls are average/likely to be upper limit.

flip1995 commented on tests/ui/test_without_fail_case.rs:
```diff
@@ -0,0 +1,99 @@
+    let b = a[0];
+}
+
+fn implicit_panic() {
+    panic!("this is an implicit panic");
+}
```
I would expect that most of those functions, that move the assertion or panic out of the `#[test]` function is defined inside the same module as the tested code. Pruning the functions this lint checks with this heuristic, whether they panic, might prevent big performance impacts this lint might have.

flip1995 commented on clippy_lints/src/test_without_fail_case.rs:
```diff
@@ -0,0 +1,268 @@
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'tcx>) {
+        // Only interested in functions that are annotated with `#[test]`.
+        if let ItemKind::Fn(_, _, body_id) = item.kind
+            && is_in_test_function(cx.tcx, item.hir_id())
```
This lint only runs on tests. This also means, that the lint can only detect something, if Clippy is run with `cargo clippy --all-targets` or `cargo clippy --tests`. This should be mentioned in the documentation of the lint.

I'm not sure, if `lintcheck` is run with `--all-targets`. It at least didn't trigger once for all the crates we test on.

bors commented:
:umbrella: The latest upstream changes (presumably #13376) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

kayagokalp commented:
Hey couple of life stuff happened and I wasn't able to look into this the past couple of days. 

I'll be finishing this and addressing reviews next week,  thanks!
